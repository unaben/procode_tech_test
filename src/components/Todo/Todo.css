.todo-wrapper {
  background-color: #353434;
  padding: 2%;
  max-width: fit-content;
  margin: 3% auto;
  max-height: 80vh;
  overflow-y: auto;
  box-shadow: 0px 5px 7px rgb(27, 27, 27);
}

.todo-content {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

@media only screen and (max-width: 768px) {
  .todo-wrapper {
    max-width: 80%;
    padding-left: 1em;
  }
}
