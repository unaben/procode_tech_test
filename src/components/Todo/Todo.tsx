import { useState } from "react";
import FormInput from "../FormInput/FormInput";
import TodoListBtn from "../TodoListBtn/TodoListBtn";
import TodoList from "../TodoList/TodoList";
import { useLocalStorage } from "../hook/useLocalStorage";
import { ICompletedTodo } from "../../models/TodoModel";
import { useAppDispatch, useAppSelector } from "../../app/hooks";
import { deleteTodo, getAllTodos } from "../../features/todo/TodoSlice";
import { executeDelete, generateTime } from "../../helper/helper";
import SearchTodo from "../SearchTodo/SearchTodo";
import "./Todo.css";
import { message } from "../../translation/ENT";
import useSearch from "../hook/useSearch";

const Todo = () => {
  const [searchTerm, setSearchTerm] = useState<string>("");
  const [toggleTodoBtn, setToggleTodoBtn] = useState<boolean>(false);
  const [completeTodos, setCompleteTodos] = useLocalStorage<ICompletedTodo[]>(
    "completedTodos",
    []
  );
  const allTodos = useAppSelector(getAllTodos);
  const { searchData } = useSearch(allTodos, searchTerm);
  const dispatch = useAppDispatch();

  const handleDelete = (id: number) => {
    dispatch(deleteTodo(id));
  };

  const handleDeleteCompletedTodo = (id: number) => {
    const newValue = executeDelete(id, completeTodos);
    setCompleteTodos(newValue);
  };

  const checkCompleteTodo = (id: number) => {
    const completedAt = generateTime();
    const filteredItem = { ...allTodos[id], completedAt };
    setCompleteTodos([...completeTodos, filteredItem]);
    handleDelete(id);
  };
  return (
    <div className="todo-wrapper">
      <FormInput setToggleTodoBtn={setToggleTodoBtn} />
      <div className="todo-content">
        <TodoListBtn
          toggleTodoBtn={toggleTodoBtn}
          setToggleTodoBtn={setToggleTodoBtn}
        />
        <p>
          {toggleTodoBtn === false
            ? message.totalPending
            : message.totalCompleted}
          : {toggleTodoBtn === false ? allTodos.length : completeTodos.length}
        </p>
        <SearchTodo searchTerm={searchTerm} setSearchTerm={setSearchTerm} />
      </div>
      <div className="todo-list">
        {toggleTodoBtn === false ? (
          <TodoList
            data={searchData}
            onComplete={checkCompleteTodo}
            onDelete={handleDelete}
            setToggleTodoBtn={setToggleTodoBtn}
          />
        ) : (
          <TodoList
            data={completeTodos}
            onDelete={handleDeleteCompletedTodo}
            onComplete={() => {}}
            setToggleTodoBtn={() => {}}
          />
        )}
      </div>
    </div>
  );
};

export default Todo;
