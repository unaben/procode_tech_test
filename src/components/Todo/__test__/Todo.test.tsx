import { render, screen } from "@testing-library/react";
import Todo from "../Todo";
import { Provider } from "react-redux";
import { store } from "../../../app/store";

describe("Todo component", () => {
  it("should renders correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <Todo />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should renders pragraph text correctly", () => {
    render(
      <Provider store={store}>
        <Todo />
      </Provider>
    );
    const paragraphText = screen.getByText(/Pending Todos/i);
    expect(paragraphText).toBeInTheDocument();
  });
});
