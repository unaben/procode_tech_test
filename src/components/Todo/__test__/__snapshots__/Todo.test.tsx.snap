// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo component should renders correctly 1`] = `
<div>
  <div
    class="todo-wrapper"
  >
    <form
      class="todo-input"
    >
      <div
        class="todo-input-item"
      >
        <label>
          Title
        </label>
        <input
          id="title"
          name="title"
          placeholder="Enter todo title..."
          type="text"
          value=""
        />
      </div>
      <div
        class="todo-input-item"
      >
        <label
          class="description"
        >
          Description
        </label>
        <input
          id="description"
          name="description"
          placeholder="Enter description...."
          type="text"
          value=""
        />
      </div>
      <div
        class="todo-input-item"
      >
        <button
          class="primary-btn"
          type="submit"
        >
          Add
        </button>
      </div>
    </form>
    <div
      class="todo-content"
    >
      <div
        class="btn-area"
      >
        <button
          class="secondary-btn active"
        >
          Pending
        </button>
        <button
          class="secondary-btn false"
        >
          Completed
        </button>
      </div>
      <p>
        Pending Todo
        : 
        0
      </p>
      <form
        class="todo-input-item"
      >
        <input
          id="search-term"
          name="searchTerm"
          placeholder="Search todos...."
          type="text"
          value=""
        />
      </form>
    </div>
    <div
      class="todo-list"
    >
      <div />
    </div>
  </div>
</div>
`;
